src/runtime-rs/crates/resource/src/network/endpoint/physical_endpoint.rs

```diff
@@ -0,0 +1,140 @@
+use hypervisor::{device, Hypervisor};
+
+use super::Endpoint;
+use crate::network::utils::{self, link};
+
+pub const SYS_PCI_DEVICES_PATH: &str = "/sys/bus/pci/devices";
+
+#[derive(Debug)]
+pub struct VendorDevice {
+    vendor_id: String,
+    device_id: String,
+}
+
+impl VendorDevice {
+    pub fn new(vendor_id: &str, device_id: &str) -> Self {
+        Self {
+            vendor_id: vendor_id.to_string(),
+            device_id: device_id.to_string(),
+        }
+    }
+
+    pub fn vendor_device_id(&self) -> String {
+        let vendor_device_id = format!("{} {}", &self.vendor_id, &self.device_id);
+        String::from(vendor_device_id.trim())
+    }
+}
+
+#[derive(Debug)]
+pub struct PhysicalEndpoint {
+    iface_name: String,
+    hard_addr: String,
+    bdf: String,
+    driver: String,
+    vendor_device_id: VendorDevice,
+}
+
+impl PhysicalEndpoint {
+    pub fn new(name: &str, hardware_addr: &[u8]) -> Result<Self> {
+        let driver_info = link::get_driver_info(name).context("get driver info")?;
+        let bdf = driver_info.bus_info;
+        let sys_pci_devices_path = Path::new(SYS_PCI_DEVICES_PATH);
+        // get driver by following symlink /sys/bus/pci/devices/$bdf/driver
+        let driver_path = sys_pci_devices_path.join(&bdf).join("driver");
+        let link = driver_path.read_link().context("read link")?;
+        let driver = link
+            .file_name()
+            .map_or(String::new(), |v| v.to_str().unwrap().to_owned());
+
+        // get vendor and device id from pci space (sys/bus/pci/devices/$bdf)
+        let iface_device_path = sys_pci_devices_path.join(&bdf).join("device");
+        let contents = std::fs::read(&iface_device_path)
+            .context(format!("read device path {:?}", &iface_device_path))?;
+        let device_id = String::from_utf8(contents).context("device id from utf8")?;
+
+        let iface_vendor_path = sys_pci_devices_path.join(&bdf).join("vendor");
+        let contents = std::fs::read(&iface_vendor_path)
+            .context(format!("read vendor path {:?}", &iface_vendor_path))?;
+        let vendor_id = String::from_utf8(contents).context("vendor id from utf8")?;
+
+        Ok(Self {
+            iface_name: name.to_string(),
+            hard_addr: utils::get_mac_addr(hardware_addr),
+            vendor_device_id: VendorDevice::new(&vendor_id, &device_id),
+            driver,
+            bdf,
+        })
+    }
+}
+
+#[async_trait]
+impl Endpoint for PhysicalEndpoint {
+    async fn name(&self) -> String {
+        self.iface_name.clone()
+    }
+
+    async fn hardware_addr(&self) -> String {
+        self.hard_addr.clone()
+    }
+
+    async fn attach(&self, hypervisor: &dyn Hypervisor) -> Result<()> {
+        // bind physical interface from host driver and bind to vfio
+        device::bind_device_to_vfio(
+            &self.bdf,
+            &self.driver,
+            &self.vendor_device_id.vendor_device_id(),
+        )
+        .context(format!(
```

Comment:
with_context is better, because it is evaluated lazily.

Reply:
Thanks for your review, since the `context` has been used everywhere, I'll raise an issue(#4129) to change this to `with_context`.